rfcs/ion_templates.md

```diff
@@ -0,0 +1,472 @@
+    hasAirbags: true
+  }]
+}
+```
+
+Template definitions can refer to other templates, but only those templates that have already been defined (i.e. templates with a lower ID) in order to cheaply avoid cycles.
+
+
+## Invoking a template
+
+### Invocation syntax
+
+```js
+// Invoke template 0 with the following parameters
+@0("Toyota", "Camry", 2017)
+```
+
+Parameters to a template invocation can themselves be template invocations. There is no restriction on which templates can be invoked as parameters.
+
+### Expanded data
+
+```js
+{ 
+  make: "Toyota",
+  model: "Camry",
+  year: 2017,
+  frame: "sedan",
+  numberOfWheels: 4,
+  transmission: "automatic",
+  hasAirbags: true
+}
+```
+
+In binary Ion, the template invocation requires 18 bytes to encode while the expanded data would require 42 bytes.
+
+## Binary layout
+
+See the [Ion 1.0 binary encoding spec](http://amzn.github.io/ion-docs/docs/binary.html#typed-value-formats) for a detailed explanation of the terms used in the diagram below.
+
+```
+            7       4 3       0
+           +---------+---------+
+Template   |   15    |    L    |
+           +---------+---------+======+
+           :     length [VarUInt]     :
+           +--------------------------+
+           |     template ID [VarUInt]|
+           +--------------------------+
+           :           value          :
+           +==========================+
+                         ⋮
+```
+
+* `L` values of `0` and `null` are available to be used as sentinels. 
+* This document treats `L`=0 as a template blank: `@_`
+* `L`=14 is a variable `length` encoding.
+* `L`=15 is reserved for future specs to accommodate not-yet-designed template invocation methods. (e.g. “splicing”)
+* Templates have a minimum length of 2 (1 byte type descriptor, 1 byte template ID)
+* A template blank (`L`=0, `@_`) found outside of a template definition or template invocation is illegal and should be treated as an error.
+* Template invocations with an ID that cannot be resolved are illegal and should be treated as an error. 
+    * (This differs from `symbol`s with an ID that cannot be resolved. `symbol`s are values while `template`s are an encoding detail. If a `Reader` encounters an unresolvable `symbol`, it can still tell the user that it is positioned over a symbol and even provide the unresolvable ID number.)
+
+
+
+## Text encoding within s-expressions
+
+Because `@` is a valid operator character within an s-expression, template blanks in the text format must be annotated with the special `$ion_template` annotation. Per the spec, symbols starting with `$ion` are reserved and may not be used in user data.
+
+### Blanks in a template definition s-expression
+
+```js
+$ion_template_table::{
+  templates : [{ // template 0
+    ($ion_template::@_)  
+  }]
+}  
+```
+
+### Template invocation inside an s-expression
+
+```js
+// Invoke template 0 from within an s-expression.
+($ion_template::@0(123))
+// ((123))
+```
```

Comment:
I *think* this has some lexical implications that really need to be spelled out.

To wit, `@` is an *operator* token on its own, `_` is an identifier token on its own, `0` is an integer token on its own, so does this mean that any or all of the following are equivalent?

```js
($ion_template::@    _)
($ion_template::'@' /*ignore me!*/ '_')
($ion_template::'@_')
```
and

```js
($ion_template::@  0)
($ion_template::'@'  '0')
($ion_template::'@0')
```

Reply:
Another idea is to annotate s-expressions to indicate that the s-expression that follows is Ion Template related syntax. Here is a rough idea 

Template invocations 

`$ion_template::(1  a b c)` 

Borrowing from the way function calls are typically encoded in an s-expression, first element is the template ID the remaining elements are the parameters. 

`$ion_template::(1 $ion_template(2 a) c)`  an example with template invocation as a parameter. 

Template definition 
```
$ion_template_table::{
  templates : [{ // template 0
    $ion_template::(_)  // drops the `@` 
  }]
}  
```

Taking the idea a little further, the syntax for templates in Ion text (outside of s-expressions) could also be moved to this style, e.g. 

```
$ion_template_table::{
  templates : $ion_template[{ // template 0
    make: @_,
    model: @_,
    year: @_,
    frame: sedan,
    numberOfWheels: 4,
    transmission: automatic,
    hasAirbags: true
  }]
}
```
Given that the context is now specified through an annotation we get some more freedom on the syntax, for example I would drop the `@` signs in the above example and have `_` only. e.g., 

```
$ion_template_table::{
  templates : $ion_template[{ // template 0
    make: _,
    model: _,
    year: _,
    frame: sedan,
    numberOfWheels: 4,
    transmission: automatic,
    hasAirbags: true
  }]
}
```

Personally I would have one annotation for template definition ( `$ion_template`) and another for template invocation (`$ion_template_call`) but that might be overkill.  I feel like I missing something here--some property that makes the annotations only idea not work--but I cannot put my finger on it.